Add Xfs package to Bottlerocket

yeazelm opened the pull request:
**Issue number:** 3036

**Description of changes:**
This adds xfsprogs and it's dependencies to Bottlerocket. There is more work to properly make the variant boundary work but I have a rough set of changes to force the data partition into `xfs`. 

Note: We may hold back the aws-dev variant until we feel we are ready but for the time being, it simply makes the tools available, the data partition still remains `ext4` for the time being.


**Testing done:**
I build my own variant with xfs hard coded into the data partition. It boots, resizes, and you can see xfs data with xfs_* tools:
```
bash-5.1# xfs_info /local
meta-data=/dev/nvme1n1p1         isize=512    agcount=401, agsize=130816 blks
         =                       sectsz=4096  attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=0
         =                       reflink=1    bigtime=1 inobtcount=1
data     =                       bsize=4096   blocks=52428288, imaxpct=25
         =                       sunit=128    swidth=128 blks
naming   =version 2              bsize=16384  ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=16384, version=2
         =                       sectsz=4096  sunit=1 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
```



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
packages: add libinih

inih is a library needed for programs like xfsprogs and provides a
small dependency for ini parsing.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
packages: add liburcu

liburcu aka userspace-rcu is used for xfsprogs. This provides the libraries
needed to work with xfsprogs.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
packages: add xfsprogs

Adding xfsprogs for manipulation of XFS filesystems on Bottlerocket.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

bcressey commented on packages/libinih/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+releases-url = "https://github.com/benhoyt/inih/archive/refs/tags/"
+
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/benhoyt/inih/archive/refs/tags/r56.tar.gz"
```
```suggestion
url = "https://github.com/benhoyt/inih/archive/refs/tags/r56.tar.gz"
path = "inih-r56.tar.gz"
```

bcressey commented on packages/libinih/libinih.spec:
```diff
@@ -0,0 +1,48 @@
+Summary: Simple INI file parser library
+License: BSD
+URL: https://github.com/benhoyt/inih
+Source0: https://github.com/benhoyt/inih/archive/refs/tags/r%{version}.tar.gz
```
```suggestion
Source0: https://github.com/benhoyt/inih/archive/refs/tags/r%{version}.tar.gz#/inih-%{version}.tar.gz
```

bcressey commented on packages/libinih/libinih.spec:
```diff
@@ -0,0 +1,48 @@
+%{_cross_includedir}/ini.h
+
+
+%changelog
```
nit:
```suggestion
%files devel
%{_cross_pkgconfigdir}/inih.pc
%{_cross_libdir}/libinih.so
%{_cross_includedir}/ini.h

%changelog
```

bcressey commented on packages/libuserspace_rcu/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "libuserspace_rcu"
```
I would prefer to call this `liburcu` here and in the spec, since that matches the library name.

yeazelm replied:
Updated the naming.

bcressey commented on packages/libuserspace_rcu/libuserspace_rcu.spec:
```diff
@@ -0,0 +1,67 @@
+
+%package devel
+Summary: Development package for %{name}
+Requires: %{name}
```
nit: the other packages follow these conventions for summaries -
```
main -
Summary: Library for <X>

devel -
Summary: Files for development using the library for <X>
```

So I would do this as:
```
Summary: Library for userspace RCU
Summary: Files for development using the library for userspace RCU
```

This is pedantic but is meant to reduce any effort spent on creative writing for these fields. Same rationale applies for `%description`.

yeazelm replied:
Makes sense to me! Corrected!

yeazelm replied:
I missed this update for libinih and will update shortly.

bcressey commented on packages/libuserspace_rcu/libuserspace_rcu.spec:
```diff
@@ -0,0 +1,67 @@
+%exclude %{_cross_datadir}/doc/*
+
+%files devel
+%{_cross_includedir}/*
```
`%{_cross_includedir}/*` and `%{_cross_includedir}/*.h` below seem redundant.

yeazelm replied:
Fixed

bcressey commented on packages/libuserspace_rcu/libuserspace_rcu.spec:
```diff
@@ -0,0 +1,67 @@
+%{_cross_libdir}/liburcu-common.so.8*
+%{_cross_libdir}/liburcu-mb.so.8*
+%{_cross_libdir}/liburcu-memb.so.8*
+%{_cross_libdir}/liburcu.so.8*
```
Can we just include this version of the library, or whichever one xfsprogs uses? I don't want to ship a bunch of different implementations when we only use one.

yeazelm replied:
I was able to determine xfsprogs only links against `liburcu.so.8` and `liburcu-common.so.8`

bcressey commented on packages/xfsprogs/0001-libxfs-do-not-try-to-run-the-crc32selftest.patch:
```diff
@@ -0,0 +1,44 @@
+From 353e3b5a88849efced5b9faedbf8b64c6a183933 Mon Sep 17 00:00:00 2001
+From: Matthew Yeazel <yeazelm@amazon.com>
+Date: Tue, 13 Jun 2023 20:06:06 +0000
+Subject: [PATCH] [PATCH] libxfs: do not try to run the crc32selftest
```
Let's keep the original buildroot patch header for attribution purposes.

yeazelm replied:
Whoops, I meant to do this and spaced it, thanks for the catch!

bcressey commented on packages/xfsprogs/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+releases-url = "https://xfs.wiki.kernel.org"
+
+[[package.metadata.build-package.external-files]]
+url = "http://kernel.org/pub/linux/utils/fs/xfs/xfsprogs/xfsprogs-5.15.0.tar.xz"
```
Does the latest version not work? Typically these packages are moved to the latest version as part of the update cycle, so if it doesn't work now then it'll just break for the next maintainer that comes along.

yeazelm replied:
I was having difficulties compiling with some strange errors. I've lost them since but will try moving to 6.3.0 and see if they reappear.

yeazelm replied:
I moved back to 6.3.0 and it builds and worked perfectly. I think it might have been other things stabilizing with liburcu so it seems fine now.

bcressey commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,86 @@
+
+%build
+%cross_configure \
+  --enable-editline=yes \
+  --enable-blkid=yes \
+  --enable-lto=no \
+  --enable-editline=no \
+  --enable-scrub=no
```
```suggestion
  --enable-blkid=yes \
  --enable-lto=no \
  --enable-editline=no \
  --enable-scrub=no
```

bcressey commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,86 @@
+  --enable-scrub=no
+
+
+%make_build \
+  INCLUDEDIR=%{_cross_includedir} \
+  LIBDIR=%{_cross_libdir}
```
Are `INCLUDEDIR` and `LIBDIR` not set to the correct default values?

yeazelm replied:
These were lingering from other failures, they do not add anything.

bcressey commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,86 @@
+
+%install
+make DIST_ROOT=%{buildroot} install install-dev \
+	PKG_ROOT_SBIN_DIR=%{_cross_sbindir} PKG_ROOT_LIB_DIR=%{_cross_libdir}
```
No hard tabs!

yeazelm replied:
Not sure how it came in, sorry!

bcressey commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,86 @@
+%exclude %{_cross_mandir}
+%exclude %{_cross_datadir}/doc/*
+%exclude %{_cross_datadir}/locale/*
+%{_cross_sbindir}/*
```
Can we list out these programs?

yeazelm replied:
Listed them out instead.

bcressey commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,86 @@
+%install
+make DIST_ROOT=%{buildroot} install install-dev \
+	PKG_ROOT_SBIN_DIR=%{_cross_sbindir} PKG_ROOT_LIB_DIR=%{_cross_libdir}
+
+  	
```
nit: I would clean up some of this double-spacing

etungsten commented on packages/libinih/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+path = "/dev/null"
+
+[package.metadata.build-package]
+releases-url = "https://github.com/benhoyt/inih/archive/refs/tags/"
```
This link 404s for me.
```suggestion
releases-url = "https://github.com/benhoyt/inih/releases"
```

yeazelm replied:
Good call out. I can fix this.

etungsten commented on packages/liburcu/liburcu.spec:
```diff
@@ -0,0 +1,67 @@
+Name: %{_cross_os}liburcu
+Version: 0.14.0
+Release: 1%{?dist}
+Summary: Library for userspace RCU
+License: LGPL-2.1-or-later
+URL: http://liburcu.org
+Source0: http://lttng.org/files/urcu/userspace-rcu-latest-0.14.tar.bz2
```
nit:
```suggestion
Version: 0.14
Release: 1%{?dist}
Summary: Library for userspace RCU
License: LGPL-2.1-or-later
URL: http://liburcu.org
Source0: http://lttng.org/files/urcu/userspace-rcu-latest-%{version}.tar.bz2
```

yeazelm replied:
The version is nonstandard, the version *is* 0.14.0 but the tarball that is downloaded is named 0.14. So to use `%{version}` during the the `autosetup` step, it needs to have the .0 in it. So either we have to do conversion in the `Source0` or the `autosetup` and I felt like it is "more correct" to change the name from the Source than in the `autosetup`. I can switch it around though if you feel its better this way.

etungsten replied:
The way you currently have it is fine. I wasn't aware of how the version is used elsewhere.

etungsten commented on packages/xfsprogs/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+path = "/dev/null"
+
+[package.metadata.build-package]
+releases-url = "https://xfs.wiki.kernel.org"
```
Would this be more accurate?
```suggestion
releases-url = "https://mirrors.edge.kernel.org/pub/linux/utils/fs/xfs/xfsprogs/"
```

yeazelm replied:
Yeah, that is probably more accurate. I was thinking of this more as the `URL` in the specfile but strictly speaking the releases are found at that URL.

stmcginnis commented on variants/aws-dev/Cargo.toml:
```diff
@@ -36,6 +36,7 @@ included-packages = [
     "strace",
     "tcpdump",
     "chrony-tools",
+    "xfsprogs",
```
Looks good for now while working through next steps. Just to note, would be good to get this moved to be part of the `os` package once we are ready.

yeazelm replied:
Correct, for the moment, this is just to ensure it gets built at all. The next PR will be adding in the proper support for XFS in variants which will move this to the right place.

arnaldo2792 replied:
I agree we should only include it in the variants that will be released with XFS. Existing variants should remain in ext4.

stmcginnis reviewed (approved):
lgtm

arnaldo2792 commented on packages/liburcu/liburcu.spec:
```diff
@@ -0,0 +1,67 @@
+%exclude %{_cross_libdir}/liburcu-qsbr.so
+%exclude %{_cross_libdir}/liburcu-signal.so
+
+%changelog
```
nit: you don't need this line

arnaldo2792 commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,93 @@
+Version: 6.3.0
+Release: 1%{?dist}
+Summary: Utilities for managing the XFS filesystem
+License: GPL+ and LGPLv2+
```
This package includes copies of the GPL-2.0-only and LGPL-2.1-only licenses, this should match those licenses.

yeazelm replied:
I'll take a look and make sure this is right.

arnaldo2792 commented on packages/liburcu/liburcu.spec:
```diff
@@ -0,0 +1,67 @@
+Version: 0.14.0
+Release: 1%{?dist}
+Summary: Library for userspace RCU
+License: LGPL-2.1-or-later
```
Different items in this package are distributed under different licenses, according to the LICENSE file in the sources:

```
The library part is distributed under LGPLv2.1 or later.
.....
.....
xchg() primitive has been rewritten from scratch starting from atomic_ops 1.2
which has a MIT-style license that is intended to allow use in both free and
proprietary software:
....
....
Library test code is distributed under the GPLv2 license. See gpl-2.0.txt for
...
...
The following build-related macro is under GPLv3 (or later):
```

yeazelm replied:
Same with this, I will take a look and make sure its correct.

arnaldo2792 commented on packages/liburcu/liburcu.spec:
```diff
@@ -0,0 +1,67 @@
+find %{buildroot} -type f -name "*.a" -delete
+
+%files
+%license LICENSE
```
The sources distribute copies of the licenses that apply to the package, should you use those instead of the `LICENSE`file?

arnaldo2792 commented on packages/libinih/libinih.spec:
```diff
@@ -0,0 +1,46 @@
+Version: 56
+Release: 1%{?dist}
+Summary: Simple INI file parser library
+License: BSD
```
This package is released under the new BSD-3 clause:

```bash
inih-r56 is 📦 v56 ❯ askalono crawl .
./LICENSE.txt
License: BSD-3-Clause (original text)
```

arnaldo2792 commented on packages/libinih/libinih.spec:
```diff
@@ -0,0 +1,46 @@
+%{_cross_libdir}/libinih.so
+%{_cross_includedir}/ini.h
+
+%changelog
```
This line isn't needed

arnaldo2792 commented on variants/Cargo.lock:
```diff
@@ -630,13 +630,6 @@ dependencies = [
 ]
 
-[[package]]
-name = "libinih"
```
nit: some cleanup in this commit.

yeazelm replied:
Yeah, I'll do this cleanup if I can, I've tried twice in rebases to fix it but I'll take another shot at it.

arnaldo2792 commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,93 @@
+%{_cross_includedir}/xfs/xqm.h
+%{_cross_libdir}/*.so
+
+%changelog
```
This line isn't needed.

yeazelm commented on Dockerfile:
```diff
@@ -102,6 +103,7 @@ RUN rpmdev-setuptree \
    && echo -e -n "${GRUB_SET_PRIVATE_VAR:+%bcond_without grub_set_private_var\n}" >> .bconds \
    && echo -e -n "${SYSTEMD_NETWORKD:+%bcond_without systemd_networkd\n}" >> .bconds \
    && echo -e -n "${UNIFIED_CGROUP_HIERARCHY:+%bcond_without unified_cgroup_hierarchy\n}" >> .bconds \
+   && echo -e -n "${XFS_DATA_PARTITION:+%bcond_without xfs_data_partition\n}" >> .bconds \
```
I added this for if we need to do conditional specfile things. I didn't end up adding `xfsprogs` conditionally to the image based upon this but we could if we want. If we don't see the need for this I can remove it as well.

arnaldo2792 replied:
Should it be something more generic? Like `DATA_PARTITION_FORMAT=<the-format>`? Wouldn't OOTB benefit from this? Say I'm a weird developer that wants to use btrfs for my data partition in my OOT variant.

bcressey replied:
> I didn't end up adding xfsprogs conditionally to the image based upon this but we could if we want.

It's fine to add the bcond support here even if we aren't using it, just for consistency. I did the same in #3097.

yeazelm replied:
I'm having a hard time replying to the other threads about @arnaldo2792's comment:
>Should it be something more generic? Like DATA_PARTITION_FORMAT=<the-format>? Wouldn't OOTB benefit from this? Say I'm a weird developer that wants to use btrfs for my data partition in my OOT variant.

I did look at trying to make it work this way and discovered what @bcressey called out in his comment about `image-layout`. Since `image-features` are booleans these lend themselves as something like `XFS_DATA_PARTITION` vs `DATA_PARTITION_FILESYSTEM` or similar. As @bcressey mentioned, we probably would need to shift to using either `image-layout` (which I feel is a little awkward) or adding in yet another data structure for this. I actually like the idea of leaving the door open for other filesystems in the future but also feel like it might be over-indexing on something we have no concrete plans to do. I can shift to that approach if we think it's useful for OOTB work ensuring we have this flexibility in the future. Ultimately though, my opinion is that there is a lot of other work involved in adding something like `btrfs` and adding in another filesystem would be the right time to solve this more expansively but I could be overruled if that doesn't align with other features.

yeazelm commented on tools/rpm2img:
```diff
@@ -408,8 +417,28 @@ mkfs_data() {
+    # Create a file to write the filesystem to first
+    dd if=/dev/zero of="${BOTTLEROCKET_DATA}" bs=1M count=${size%?}
+    # block size of 4096, directory block size of 16384
+    # enable inotbtcount and bigtime
```
We could also enable reflink if we want. I haven't done any testing to validate if it changes performance at all for Bottlerocket.

yeazelm replied:
I noticed that it was enabled even without specifying it (which is fine, it only comes into play if the tool making the changes on the filesystem is asking for it) so I decided to add it explicitly in anyway.
```bash
bash-5.1# xfs_spaceman -p xfs_info -c info /local
meta-data=/dev/nvme1n1p1         isize=512    agcount=41, agsize=130816 blks
         =                       sectsz=4096  attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=0
         =                       reflink=1    bigtime=1 inobtcount=1 nrext64=0
data     =                       bsize=4096   blocks=5242368, imaxpct=25
         =                       sunit=128    swidth=128 blks
naming   =version 2              bsize=16384  ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=16384, version=2
         =                       sectsz=4096  sunit=1 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
```

arnaldo2792 commented on packages/liburcu/liburcu.spec:
```diff
@@ -0,0 +1,65 @@
+%autosetup -n userspace-rcu-%{version}
+
+%build
+%cross_configure
+
+# "fix" rpath
+sed -i 's|^hardcode_libdir_flag_spec=.*|hardcode_libdir_flag_spec=""|g' libtool
+sed -i 's|^runpath_var=LD_RUN_PATH|runpath_var=DIE_RPATH_DIE|g' libtool
```
I'm curious why is this needed. Could you please elaborate on why this replacement is needed? Should we do what fedora does?

```spec
# Reinitialize libtool with the fedora version to remove Rpath
autoreconf -vif -W all,error
%configure --disable-static # You could include --disable-static and save yourself the find command below
```

yeazelm replied:
From what I understand, we do this fix in a few other autoconf packages (and it works quite well for this package as well). I believe it was introduced in https://github.com/bottlerocket-os/bottlerocket/pull/92 and is there to ensure the rpaths are sane and makes `check-buildroot`|`check-rpaths` happy.

> Should we do what fedora does?

I didn't play with `autoreconf` much but it starts breaking the build in new ways and still doesn't fix the rpaths. Nonetheless the --disable-static might be useful.

arnaldo2792 commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,93 @@
+make DIST_ROOT=%{buildroot} install install-dev \
+  PKG_ROOT_SBIN_DIR=%{_cross_sbindir} PKG_ROOT_LIB_DIR=%{_cross_libdir}
+
+rm -f %{buildroot}/%{_cross_libdir}/*.{la,a}
```
Same as elsewhere, is there a flag you can use to disable compiling static libraries?

yeazelm replied:
I didn't see a specific flag for it for `xfsprogs`. Nonetheless, is the approach Fedora takes.

arnaldo2792 commented on packages/release/prepare-local-fs.service:
```diff
@@ -10,8 +10,9 @@ RefuseManualStop=true
 [Service]
 Type=oneshot
 
+EnvironmentFile=/usr/lib/data-partition-filesystem
```
Instead of an environment variable, what if `prepare-local-fs.service` becomes an `in` file, and in the `ExecStart` directive you use `DATA_PARTITION_FILESYSTEM` to replace it with `sed`?

yeazelm replied:
See @bcressey's comment below but I think I'll convert to his approach where its an `image-features.env` and make this a bit more generic for any other features. This allows them to be a bit more consistent. I think I like the `EnvironmentFile` approach a bit more. Is there a reason you would prefer the `in` file and `sed`?

bcressey commented on packages/libinih/libinih.spec:
```diff
@@ -0,0 +1,44 @@
+%{summary}.
+
+%package devel
+Summary: Files for development using the Simple INI file parser library
```
nit:
```suggestion
Summary: Files for development using the simple INI file parser library
```

bcressey commented on packages/liburcu/liburcu.spec:
```diff
@@ -0,0 +1,65 @@
+%{_cross_includedir}/*
+%{_cross_libdir}/liburcu-common.so
+%{_cross_libdir}/liburcu.so
+%{_cross_libdir}/pkgconfig/liburcu*.pc
```
Do these wildcards include headers and `.pc` files for the other liburcu libraries? It'd be good to exclude those.

yeazelm replied:
Good catch! I'll exclude.

bcressey commented on packages/release/prepare-local-fs.service:
```diff
@@ -10,8 +10,9 @@ RefuseManualStop=true
 [Service]
 Type=oneshot
 
+EnvironmentFile=/usr/lib/data-partition-filesystem
```
nit: should use `/usr/share` for non-executable content, and this could have a more generic name to entice us to extend it in the future with other feature flag env variables:
```suggestion
EnvironmentFile=/usr/share/bottlerocket/image-features.env
```

yeazelm replied:
I debated back and forth on `/usr/lib/` and `/usr/share` and apparently debated myself into the wrong place :) I'll fix the path and I really like the more generic option for these variables. Thanks for the suggestion!

bcressey commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,91 @@
+%{_cross_libdir}/*.so.*
+%{_cross_sbindir}/fsck.xfs
+%{_cross_sbindir}/mkfs.xfs
+%{_cross_sbindir}/xfs_admin
```
Need to exclude all shell scripts.

yeazelm replied:
Yeah, that is a good point. I meant to call out that we will likely want to replace functionality for, at minimum, `xfs_admin`, `xfs_info`, and `fsck.xfs` with a something written in Rust. I'll move these to excluded for now and we can have another change to add those non-shell script versions into the package.

bcressey commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,91 @@
+%{_cross_datadir}/xfsprogs/mkfs/*.conf
+%exclude %{_cross_mandir}
+%exclude %{_cross_datadir}/doc/*
+%exclude %{_cross_datadir}/locale/*
```
nit: if you define `PKG_DATA_DIR ` as `%{_cross_datadir}` then you should be able to exclude the normal paths - `%{_cross_localedir}` and `%{_cross_docdir}`.

bcressey commented on packages/xfsprogs/xfsprogs.spec:
```diff
@@ -0,0 +1,91 @@
+%{_cross_includedir}/xfs/xfs_format.h
+%{_cross_includedir}/xfs/xfs_da_format.h
+%{_cross_includedir}/xfs/xfs_log_format.h
+%{_cross_includedir}/xfs/xqm.h
```
nit: this is fine too --
```suggestion
%dir %{_cross_includedir}/xfs
%{_cross_includedir}/xfs/*.h
```

bcressey commented on packages/xfsprogs/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+
+# RPM Requires
+[dependencies]
+# none - will likely change
```
nit: can just remove this section, or just say "none" in the comment.

Mostly I use these as follows:
* `[build-dependencies]` corresponds to `BuildRequires` - if this isn't correct then the build may fail
* `[dependencies]` corresponds to `Requires` - but if it's already listed as a build dependency, we don't need to list it again

yeazelm replied:
I really assumed I would need to change this but missed that the build dependency would solve this. I'll clean up the comment.

bcressey commented on tools/buildsys/src/manifest.rs:
````diff
@@ -214,6 +214,12 @@ line arguments set in the boot configuration.
+default will remain ext4 and xfs is opt-in.
+```ignore
+[package.metadata.build-variant.image-features]
+xfs-data-partition = true
````
nit: @webern suggested this in the Secure Boot PR. I think it's for consistency with GitHub markdown.

```suggestion
`xfs-data-partition` changes the filesystem for the data partition from ext4 to xfs. The
default will remain ext4 and xfs is opt-in.

```ignore
[package.metadata.build-variant.image-features]
xfs-data-partition = true
```

bcressey commented on tools/rpm2img:
```diff
@@ -408,8 +417,28 @@ mkfs_data() {
+    # block size of 4096, directory block size of 16384
+    # enable inotbtcount and bigtime
+    # use an internal log with starting size of 64m
+    # use the mininmal 2 Allocation groups, this still overprovisions when expanded
```
```suggestion
    # use the minimal 2 Allocation groups, this still overprovisions when expanded
```

bcressey commented:
I can't seem to respond to [@arnaldo2792 's question](https://github.com/bottlerocket-os/bottlerocket/pull/3198#discussion_r1256300627) in the normal way:
> Should it be something more generic? Like DATA_PARTITION_FORMAT=<the-format>? Wouldn't OOTB benefit from this? Say I'm a weird developer that wants to use btrfs for my data partition in my OOT variant.

If we really want to do this, we can make `data-image-filesystem` a field in the `image-layout` struct (instead of a feature flag), and default it to "ext4".

Commit:
build: Add support for XFS DATA Partition

Add the ability to specify XFS for a variant's DATA partition. The
variants now allow xfs-data-partition to be set to enable this feature.
If not specified, the DATA partition defaults to ext4.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

bcressey commented on tools/rpm2img:
```diff
@@ -408,8 +417,28 @@ mkfs_data() {
+  if [ "${XFS_DATA_PARTITION}" == "yes" ] ; then
+    echo "writing XFS filesystem for DATA"
+    # Create a file to write the filesystem to first
+    dd if=/dev/zero of="${BOTTLEROCKET_DATA}" bs=1M count=${size%?}
```
nit: `${size}` is known to be non-empty at this point:

```suggestion
    dd if=/dev/zero of="${BOTTLEROCKET_DATA}" bs=1M count=${size}
```

yeazelm replied:
This is actually trimming the last character off the `${size}` since `dd` wants a count but the value contains an `M` for `mkfs.ext4`. So you end up with `1024` instead of `1024M`.
